data/src/nonempty.rs

```diff
@@ -0,0 +1,293 @@
+    fn is_empty(&self) -> bool;
+}
+
+impl<T> MaybeEmpty for BTreeSet<T> {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+impl<T> MaybeEmpty for HashSet<T> {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+impl<K, V> MaybeEmpty for BTreeMap<K, V> {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+impl<K, V, S> MaybeEmpty for HashMap<K, V, S> {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+/// Mutable set operations which don't decrease the size of the set.
+pub trait Set {
+    type Value;
+
+    /// Adds a value to the set.
+    ///
+    /// If the set did not have this value present, true is returned. If the set
+    /// did have this value present, false is returned, and the entry is not
+    /// updated.
+    fn insert(&mut self, value: Self::Value) -> bool;
+
+    /// Adds a value to the set, replacing the existing value, if any, that is
+    /// equal to the given one. Returns the replaced value.
+    fn replace(&mut self, value: Self::Value) -> Option<Self::Value>;
+}
+
+impl<T> Set for BTreeSet<T>
+where
+    T: Ord,
+{
+    type Value = T;
+
+    fn insert(&mut self, value: T) -> bool {
+        self.insert(value)
+    }
+
+    fn replace(&mut self, value: T) -> Option<T> {
+        self.replace(value)
+    }
+}
+
+impl<T, S> Set for HashSet<T, S>
+where
+    T: Eq + Hash,
+    S: BuildHasher,
+{
+    type Value = T;
+
+    fn insert(&mut self, value: T) -> bool {
+        self.insert(value)
+    }
+
+    fn replace(&mut self, value: T) -> Option<T> {
+        self.replace(value)
+    }
+}
+
+/// Mutable map operations which don't decrease the size of the map.
+pub trait Map {
+    type Key;
+    type Value;
+
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, None is returned. If the map
+    /// did have this key present, the value is updated, and the old value
+    /// is returned. The key is not updated, though; this matters for types
+    /// that can be == without being identical.
+    fn insert(&mut self, key: Self::Key, value: Self::Value) -> Option<Self::Value>;
+}
+
+impl<K, V> Map for BTreeMap<K, V>
+where
+    K: Ord,
+{
+    type Key = K;
+    type Value = V;
+
+    fn insert(&mut self, key: K, value: V) -> Option<V> {
+        self.insert(key, value)
+    }
+}
+
+/// Newtype wrapper around container types, which witnesses that the container
+/// contains at least one element.
+///
+/// Non-mutating methods of the underlying container are available through the
+/// [`Deref`] impl.
+///
+/// Mutating methods which either grow the container or don't change its size
+/// are provided via the [`Set`] and [`Map`] impls, respectively. Additionally,
+/// [`Extend`] is implemented if the container implements it.
+///
+/// To construct a [`NonEmpty`], use [`NonEmpty::from_maybe_empty`].
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct NonEmpty<T>(T);
+
+impl<T> NonEmpty<T>
```

Comment:
I think one thing that would be useful for using this is having some kind of `pub fn new` that takes the first element. Sometimes you know you're going to have the first element at hand and then `insert` the rest. Otherwise, we'll likely end up doing `from_maybe_empty(set).expect("set is not empty. qed.")` :upside_down_face: 

It could be in terms of a `trait Insertable` and so we would have:
```rust
pub fn new(v: T::Value) -> Self where T: Insertable + Default {
  let mut container = T::default();
  container.insert(v);
  Self(container)
}
```

Reply:
Good point. I don't see a way to overload this method, though.

Reply:
https://github.com/radicle-dev/radicle-link/pull/610/commits/b1c97d14420aef159f58e1a66ae75e3a238f1fc1

Reply:
:shrug: I guess they're more used to `expect`ing the worst :)